pkg/util/validation/validate.go

```diff
@@ -247,22 +246,24 @@ type MetadataValidationConfig interface {
 	MaxMetadataLength(userID string) int
 }
 
-// ValidateMetadata returns an err if a metric metadata is invalid.
-func ValidateMetadata(cfg MetadataValidationConfig, userID string, metadata *mimirpb.MetricMetadata) error {
+// CleanAndValidateMetadata returns an err if a metric metadata is invalid.
+func CleanAndValidateMetadata(cfg MetadataValidationConfig, userID string, metadata *mimirpb.MetricMetadata) error {
 	if cfg.EnforceMetadataMetricName(userID) && metadata.GetMetricFamilyName() == "" {
 		DiscardedMetadata.WithLabelValues(reasonMissingMetricName, userID).Inc()
 		return newMetadataMetricNameMissingError()
 	}
 
 	maxMetadataValueLength := cfg.MaxMetadataLength(userID)
+
+	if len(metadata.Help) > maxMetadataValueLength {
+		metadata.Help = metadata.Help[:maxMetadataValueLength]
```

Comment:
I've added a fix that does a for loop but still uses slicing to avoid allocating memory

Reply:
I'm not sure it's a good idea to have this loop, plus we don't check if the input it utf-8 (I don't think), so maybe it's ok to just slice the string after all?

Reply:
I'm not sure what the loops solves. A bit of searching makes me think that "runes" are the right way to deal with this. That also matches [what Prometheus does](https://github.com/prometheus/prometheus/blob/c1b669bf9b0b1286ece53c977262089119783105/tsdb/exemplar.go#L219-L256) when dealing with exemplar labels. I don't feel particularly strongly about this but it seems prone to be an issue since `HELP` text is where users are most likely to put multi-byte characters.

Reply:
`range` on strings works on runes, not bytes, so this loop finds index of last rune that can still be added to be within the size limit. It seems fine to me. I would keep it as-is, so that users don't get corrupted texts.